dm/pkg/ha/bound.go

```diff
@@ -202,46 +202,86 @@ func GetLastSourceBounds(cli *clientv3.Client) (map[string]SourceBound, int64, e
-		retryNum = defaultGetSourceBoundConfigRetry
-		sbm      map[string]SourceBound
+		bounds    map[string]SourceBound
+		newBounds map[string]SourceBound
+		configs   map[string]*config.SourceConfig
+		ok        bool
+		retryNum  = defaultGetSourceBoundConfigRetry
 	)
 	wbm, rev, err := GetSourceBound(cli, worker, source)
 	if err != nil {
-		return bound, cfg, 0, err
+		return bounds, configs, 0, err
 	}
-	if sbm, ok = wbm[worker]; !ok {
-		return bound, cfg, rev, nil
+	if bounds, ok = wbm[worker]; !ok {
+		return bounds, configs, rev, nil
+	}
+
+	appendGetUpstreamCfgOps := func(sourceBounds map[string]SourceBound, ops []clientv3.Op) []clientv3.Op {
+		for src := range sourceBounds {
+			ops = append(ops, clientv3.OpGet(common.UpstreamConfigKeyAdapter.Encode(src)))
+		}
+		return ops
+	}
+
+	getSourceCfgFromResp := func(txnResp *clientv3.TxnResponse) (map[string]*config.SourceConfig, error) {
+		if txnResp == nil || len(txnResp.Responses) < 2 {
+			return nil, nil
+		}
+		scm := make(map[string]*config.SourceConfig, 0)
+		for i := 1; i < len(txnResp.Responses); i++ {
+			cfgResp := txnResp.Responses[i].GetResponseRange()
+			scs, err1 := sourceCfgFromResp("", (*clientv3.GetResponse)(cfgResp))
+			if err1 != nil {
+				return nil, err1
+			}
+			for src, conf := range scs {
+				scm[src] = conf
+			}
+		}
+		return scm, nil
 	}
-	bound = GetSourceBoundFromMap(sbm)
 
 	for retryCnt := 1; retryCnt <= retryNum; retryCnt++ {
-		txnResp, rev2, err2 := etcdutil.DoTxnWithRepeatable(cli, etcdutil.ThenOpFunc(clientv3.OpGet(common.UpstreamBoundWorkerKeyAdapter.Encode(worker), clientv3.WithPrefix()),
-			clientv3.OpGet(common.UpstreamConfigKeyAdapter.Encode(bound.Source))))
+		ops := make([]clientv3.Op, 1, len(bounds)+1)
+		ops[0] = clientv3.OpGet(common.UpstreamBoundWorkerKeyAdapter.Encode(worker), clientv3.WithPrefix())
+		ops = appendGetUpstreamCfgOps(bounds, ops)
+		txnResp, rev2, err2 := etcdutil.DoTxnWithRepeatable(cli, etcdutil.ThenOpFunc(ops...))
 		if err2 != nil {
-			return bound, cfg, 0, err2
+			return bounds, configs, 0, err2
 		}
 
 		boundResp := txnResp.Responses[0].GetResponseRange()
 		sbm2, err2 := sourceBoundFromResp((*clientv3.GetResponse)(boundResp))
 		if err2 != nil {
-			return bound, cfg, 0, err2
+			return bounds, configs, 0, err2
 		}
-
-		newBound, ok = sbm2[worker][bound.Source]
-		// when ok is false, newBound will be empty which means bound for this worker has been deleted in this turn
-		// if bound is not empty, we should wait for another turn to make sure bound is really deleted.
-		if newBound != bound {
-			log.L().Warn("source bound has been changed, will take a retry", zap.Stringer("oldBound", bound),
-				zap.Stringer("newBound", newBound), zap.Int("retryTime", retryCnt))
+		newBounds = sbm2[worker]
+
+		// 1. newBounds and bounds are exactly the same, find source configs and return.
+		// 2. newBounds and bounds are not the same, retry, when last retry is still not the same, for loop end and return error.
```

Comment:
though bounds have been changed, bounds and configs are decoded from the same transaction, we can directly use them.

Reply:
if the bound is change, may means the source doesn't bound to this worker, i am afraid we can not use them to run source_worker. or we may no need to get `newBounds` for a check.